torch/_subclasses/fake_tensor.py

```diff
@@ -77,75 +162,9 @@ def __repr__(self):
-            return args[0].fake_device
-
-        def wrap(e, device=None):
-            if isinstance(e, torch.Tensor) and not isinstance(e, FakeTensor):
-                if device:
-                    return FakeTensor(e, device)
-                else:
-                    return FakeTensor.from_tensor(e)
-            else:
-                return e
-
-        # if we are in the dispatch mode, we will enter this function even if the inputs
-        # are not FakeTensors, and they need to be wrapped
-        if cls == FakeTensorMode:
-            args = tree_map(wrap, args)
-            kwargs = tree_map(wrap, kwargs)
-
-        # _to_copy fails when run with FakeTensors to cuda device
-        # TODO: debug
-        if func == torch.ops.aten._to_copy.default:
-            _, new_kwargs = normalize_function(
-                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
-            )
-            out_device = new_kwargs.pop("device", new_kwargs["input"].device)
-            with no_dispatch():
-                input = new_kwargs.pop("input").to("meta")
-                return FakeTensor(
-                    torch.ops.aten._to_copy(input, **new_kwargs), out_device
-                )
-
-        if _is_tensor_constructor(func):
-            assert func not in _non_kwarg_device_constructors
-            _, new_kwargs = normalize_function(
-                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
-            )
-            # cpu is default device if none is specified
-            out_device = new_kwargs.pop("device", torch.device("cpu"))
-            new_kwargs["device"] = torch.device("meta")
-            r = super().__torch_dispatch__(func, types, (), new_kwargs)
-            return FakeTensor(r, out_device)
-
-        r = super().__torch_dispatch__(func, types, args, kwargs)
-
-        # TODO: handle non-kwarg devices
-        assert func not in _device_not_kwarg_ops, f"NYI: {func}"
-
-        # if device is specified, use that
-        if kwargs.get("device", None):
-            return tree_map(partial(wrap, device=kwargs["device"]), r)
-
-        # operators which copy size from another tensor do not
-        # also take device from the size tensor
-        # other size_as operators are not builtin operators
-        if func == aten.resize_as_.default:
-            _, new_kwargs = normalize_function(
-                func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
-            )
-            # device of the input is returned
-            return tree_map(partial(wrap, device=new_kwargs["input"].device), r)
-
-        common_device = FakeTensor._find_common_device(func, args, kwargs)
-
-        return tree_map(partial(wrap, device=common_device), r)
+        def run_fn(func, types, args, kwargs):
+            return torch.Tensor.__torch_dispatch__(func, types, args, kwargs)
+        return torch_dispatch_impl(cls, func, types, args, kwargs, run_fn)
```

Comment:
@samdow and I discussed this on Thursday and we think the right way to do code reuse is you put the real implementation in the mode, and then in the subclass implementation you (1) store the mode that allocated the subclass and (2) enable that mode before redispatching on the tensors as is.

Reply:
I'm going to leave to follow up because changing this causes a few different errors (which i'll file issues / smaller repros for and/or debug) and i would like to unblock dynamo